### PR TITLE
Consolidate arrow ipc tests and increase coverage

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

re https://github.com/apache/arrow-rs/issues/3389

# Rationale for this change
 
I want to ensure that the Arrow IPC writer had sufficient test coverage before I potentially messed around with it

# What changes are included in this PR?

1. Reduce duplication in  arrow-ipc read/write tests
2. Add additional coverage
3. remove redundant arrow/ipc integration test

# Are there any user-facing changes?

No this is all test changes

### CODE REVIEW

File: arrow-integration-testing/tests/ipc_reader.rs [hunk 1]
```diff
@@ -37,51 +39,42 @@ fn read_generated_files_014() {
...
     ];
     paths.iter().for_each(|path| {
-        let file = File::open(format!(
-            "{}/arrow-ipc-stream/integration/{}/{}.arrow_file",
-            testdata, version, path
-        ))
-        .unwrap();
-
-        let mut reader = FileReader::try_new(file, None).unwrap();
+        verify_arrow_file(&testdata, version, path);
```
Comment (alamb) [inline arrow-integration-testing/tests/ipc_reader.rs:42]: I consolidated the tests so the streams and file tests are in the same function, which both reduces code duplication as well as making it easier to audit which files were included in the test

File: arrow-integration-testing/tests/ipc_reader.rs [hunk 2]
```diff
@@ -37,51 +39,42 @@ fn read_generated_files_014() {
...
 
-        // read expected JSON output
-        let arrow_json = read_gzip_json(version, path);
-        assert!(arrow_json.equals_reader(&mut reader).unwrap());
+#[test]
+fn read_0_1_7() {
+    let testdata = arrow_test_data();
+    let version = "0.17.1";
+    let paths = vec!["generated_union"];
+    paths.iter().for_each(|path| {
```
Comment (alamb) [inline arrow-integration-testing/tests/ipc_reader.rs:52]: This used to be (partially) covered in `arrow/tests/ipc.rs`

File: arrow-integration-testing/tests/ipc_reader.rs [hunk 3]
```diff
@@ -102,163 +95,119 @@ fn read_generated_be_files_should_work() {
...
 #[test]
-fn projection_should_work() {
-    // complementary to the previous test
+fn read_1_0_0_littleendian() {
     let testdata = arrow_test_data();
+    let version = "1.0.0-littleendian";
     let paths = vec![
-        "generated_interval",
         "generated_datetime",
+        "generated_custom_metadata",
```
Comment (alamb) [inline arrow-integration-testing/tests/ipc_reader.rs:111]: I was quite pleased to find that the reader/writer works for many more integration tests :bowtie:

File: arrow-integration-testing/tests/ipc_reader.rs [hunk 4]
```diff
@@ -102,163 +95,119 @@ fn read_generated_be_files_should_work() {
...
-            assert_eq!(batch.columns().len(), 1);
-            assert_eq!(datatype_0, batch.schema().fields()[0].data_type().clone());
-        });
+        verify_arrow_file(&testdata, version, path);
+        verify_arrow_stream(&testdata, version, path);
     });
 }
 
 #[test]
-fn read_generated_streams_014() {
```
Comment (alamb) [inline arrow-integration-testing/tests/ipc_reader.rs:144]: the stream tests have been incorporated into the same test as the file tests (so I didn't have to audit two separate lists of files)

File: arrow-integration-testing/tests/ipc_writer.rs [hunk 5]
```diff
@@ -40,275 +39,239 @@ fn read_and_rewrite_generated_files_014() {
...
-        let arrow_json = read_gzip_json(version, path);
-        assert!(arrow_json.equals_reader(&mut reader).unwrap());
+        roundtrip_arrow_file(&testdata, version, path);
+        roundtrip_arrow_stream(&testdata, version, path);
     });
 }
 
 #[test]
-fn read_and_rewrite_generated_streams_014() {
+fn write_0_1_7() {
```
Comment (alamb) [inline arrow-integration-testing/tests/ipc_writer.rs:48]: This used to be (partially) covered in `arrow/tests/ipc.rs`

File: arrow/tests/ipc.rs [hunk 6]
```diff
@@ -1,61 +0,0 @@
-// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (alamb) [inline arrow/tests/ipc.rs:1]: This is now completely covered by `arrow-integration-testing/tests/ipc_writer.rs ` and `arrow-integration-testing/tests/ipc_reader.rs` in a uniform manner

File: arrow-integration-testing/tests/ipc_reader.rs [hunk 7]
```diff
@@ -102,163 +95,119 @@ fn read_generated_be_files_should_work() {
         .unwrap();
 
         FileReader::try_new(file, None).unwrap();
+
+        // While the the reader doesn't error but the values are not read correctly
```
Comment (tustvold) [inline arrow-integration-testing/tests/ipc_reader.rs:99]: Should we file a ticket for this?

File: arrow-integration-testing/tests/ipc_reader.rs [hunk 8]
```diff
@@ -102,163 +95,119 @@ fn read_generated_be_files_should_work() {
...
+        "generated_decimal256",
+        "generated_dictionary",
+        "generated_dictionary_unsigned",
+        "generated_duplicate_fieldnames",
+        "generated_extension",
+        "generated_interval",
         "generated_map",
+        // fails with
+        // thread 'read_1_0_0_littleendian' panicked at 'assertion failed: `(left == right)`
+        //"generated_map_non_canonical",
```
Comment (tustvold) [inline arrow-integration-testing/tests/ipc_reader.rs:122]: I think it might be worth a ticket for this test, it seems to fail in multiple places

**Review by tustvold — APPROVED** [review]
I would recommend filing tickets for the remaining failures, but looks good to me

File: arrow-integration-testing/tests/ipc_reader.rs (hunk 7 above)
Comment (alamb) [reply arrow-integration-testing/tests/ipc_reader.rs:99]: will do

Comment (alamb) [reply arrow-integration-testing/tests/ipc_reader.rs:99]: https://github.com/apache/arrow-rs/issues/3459

File: arrow-integration-testing/tests/ipc_reader.rs (hunk 8 above)
Comment (alamb) [reply arrow-integration-testing/tests/ipc_reader.rs:122]: https://github.com/apache/arrow-rs/issues/3460

**Pull request merged**